python/cudf/cudf/core/series.py

```diff
@@ -815,6 +815,98 @@ def __array_function__(self, func, types, args, kwargs):
         else:
             return NotImplemented
 
+    def map(self, arg, na_action=None) -> "Series":
+        """
+        Map values of Series according to input correspondence.
+
+        Used for substituting each value in a Series with another value,
+        that may be derived from a function, a ``dict`` or
+        a :class:`Series`.
+
+        Parameters
+        ----------
+        arg : function, collections.abc.Mapping subclass or Series
+            Mapping correspondence.
+        na_action : {None, 'ignore'}, default None
+            If 'ignore', propagate NaN values, without passing them to the
+            mapping correspondence.
+
+        Returns
+        -------
+        Series
+            Same index as caller.
+
+        Examples
+        --------
+        >>> s = cudf.Series(['cat', 'dog', np.nan, 'rabbit'])
+        >>> s
+        0      cat
+        1      dog
+        2      NaN
+        3   rabbit
+        dtype: object
+
+        ``map`` accepts a ``dict`` or a ``Series``. Values that are not found
+        in the ``dict`` are converted to ``NaN``, unless the dict has a default
+        value (e.g. ``defaultdict``):
+
+        >>> s.map({'cat': 'kitten', 'dog': 'puppy'})
+        0   kitten
+        1    puppy
+        2      NaN
+        3      NaN
+        dtype: object
+
+        It also accepts a function:
+
+        >>> s.map('I am a {}'.format)
+        0       I am a cat
+        1       I am a dog
+        2       I am a nan
+        3    I am a rabbit
+        dtype: object
+
+        To avoid applying the function to missing values (and keep them as
+        ``NaN``) ``na_action='ignore'`` can be used:
+
+        >>> s.map('I am a {}'.format, na_action='ignore')
+        0     I am a cat
+        1     I am a dog
+        2            NaN
+        3  I am a rabbit
+        dtype: object
+        """
+        if isinstance(arg, dict):
+            result = self.copy().replace(arg)
+            for i in range(len(self)):
+                if self[i] not in arg.keys():
+                    result[i] = None
```

Comment:
This is effectively replicating the functionality of doing a `left` join, but doing it iteratively via Python. We should be able to use a join and then sort it as necessary afterwards to maintain the input order.

Reply:
Sounds good, I'll try do this with join instead

Reply:
I've updated the code but ran into some issues with just doing a left join. When I use a `left join` I get a series that contains only the replaced values but isn't filling out the rest of the DataFrame with nulls, as shown here

```
gd1 = cudf.Series(["cat", "dog", np.nan, "rabbit"])
>>> arg = {"cat": "kitten", "dog": "puppy"}
>>> print(gd1)

0       cat
1       dog
2      <NA>
3    rabbit
dtype: object

>>> Dataframe_1 = cudf.merge(cudf.Series(arg).to_frame(), gd1.to_frame(), how="left", sort=True)
>>> print(Dataframe_1 )

        0
0  kitten
1   puppy
```

I used `join` to add the nulls in after this to get the correct result. Seems to be passing all the tests. If there's another way I should be going about this let me know :D

Reply:
I think you need to swap the two frames. You need `gd1` to be the left frame and `arg` to be the right frame as you want to maintain the number of results from `gd1`.

Reply:
When I swap the two frames I get the following 

```
>>> Dataframe_1 = cudf.merge(gd1.to_frame(),cudf.Series(arg).to_frame(), how="left", sort=True)
>>> print(Dataframe_1)
        0
0     cat
1     dog
2  rabbit
3    <NA>
```

I've also tried to do this with the values replaced. I almost get the right result (order is correct) but the null values are not being added, as shown below. 

```
>>> result = gd1.replace(arg)
>>> print(result)

0    kitten
1      <NA>
2    rabbit
3     puppy
dtype: object

>>> cudf.merge(result.to_frame(),gd1.to_frame(), how="left", sort=True)

        0
0  kitten
1   puppy
2  rabbit
3    <NA>
```
We should be getting
```
0  kitten
1   puppy
2    <NA>
3    <NA>
```

I'm not sure how to add the null values in these cases without some form of iteration in python.

Reply:
My current solution gives the correct null values, but in other tests cases, the values are in the wrong order. Trying to figure out how to figure out if I can correct the order :D

Reply:
Yea, `cudf.merge` doesn't guarantee any ordering, so we'd want to have a `0 --> n` column alongside the values so that we can sort the result by it afterwards.

Reply:
Sounds good! I'll add the extra columns for sorting.